File: php/neatwi.php
<?php

/*!
 * NeatWI 1.0
 *
 * http://www.neatwi.com
 *
 * Copyright 2014 Rafal Krawczyk
 * Released under the MIT license:
 *   http://www.opensource.org/licenses/mit-license.php
 */

class NeatWI {
    
    // Source directory of controls php files
    const SOURCE_DIRECTORY = "controls";
    
    // Destination directory for js files
    const DESTINATION_DIRECTORY = "js";
    
    // Controls function names prefix
    const PREFIX = "$";
    
    // If use controls categories
    const CATEGORIES = TRUE;
    
    // New line sign in js files
    const NEW_LINE = '
';
    
    /**
     * If you use language support you have to implement this function.
     * @param string $language Language to load
     */
    private function load_language($language)
    {
        // For user implementation
    }
    
    /**
     * If you use language support you can implement this function
     * so generator can load current language after generation.
     */
    private function get_language()
    {
        // For user implementation
    }
    
    /**
     * Generate controls.
     * @param array $languages If you use language support specify languages you
     * want to generate controls for. Pass array of languages names/codes.
     * Example array('en', 'pl', 'es')
     */
    public function generate($languages = NULL)
    {
        if ($languages)
        {
            $lang = $this->get_language();
            foreach ($languages as $language)
                $this->generate_controls($language);
            if ($lang)
                $this->load_language($lang);
        }
        else
            $this->generate_controls();
    }
    
    private function generate_controls($language = NULL)
    {
        if ($language)
            $this->load_language ($language);
        
        if (self::CATEGORIES)
        {
            foreach(glob(self::SOURCE_DIRECTORY . "/*", GLOB_ONLYDIR) as $directory)
            {
                $directory_name = preg_split("#/#", $directory);
                $this->generate_directory($directory_name[1], $language);
            }
        }
        else
            $this->generate_directory(NULL, $language);
    }
    
    private function generate_directory($directory, $language)
    {
        $file_name = self::DESTINATION_DIRECTORY . '/neatwi_' . ($directory ? $directory : 'controls') . ($language ? '_' . $language : '') . '.js';
        $define_block = $controls_block = '';
        $controls_block .= '$.extend($.controls, {';
        $first = true;

        foreach(glob(self::SOURCE_DIRECTORY . ($directory ? '/' . $directory : '') . '/*.php') as $control)
        {
            if ($first)
                $first = false;
            else
                $controls_block .= ',';

            $control_name = basename($control, '.php');
            $controls_block .=  "\"" . $control_name . "\":\"";
            
            ob_start();
            include $control;
            $control_content = ob_get_clean();
            
            $script_start = array();
            $script_end = array();
            $functions_matches = array();
            
            $main_tag = array();
            if (preg_match('/< *[^>]+>/', $control_content, $main_tag, PREG_OFFSET_CAPTURE))
            {
                $attribute = array();
                if (preg_match('/[ \'"]class *= *[\'"]([^\'"]*)[\'"]/', $main_tag[0][0], $attribute, PREG_OFFSET_CAPTURE))
                {
                    $class = trim($attribute[1][0]);

                    if (sizeof($class) > 0)
                    {
                        $split = preg_split('/ /', $class);
                        if ($split[0] != $control_name)
                            $control_content = substr_replace($control_content, $control_name . ' ', $main_tag[0][1] + $attribute[1][1], 0);
                    }
                    else
                        $control_content = substr_replace($control_content, $control_name, $main_tag[0][1] + $attribute[1][1], 0);
                }
                else
                    $control_content = preg_replace('/(< *[^ >]+)/', '$1 class="' . $control_name . '"', $control_content, 1);

                preg_match('/< *[^>]+>/', $control_content, $main_tag, PREG_OFFSET_CAPTURE);
                if (preg_match('/[ \'"]id *= *[\'"]([^\'"]*)[\'"]/', $main_tag[0][0], $attribute, PREG_OFFSET_CAPTURE))
                {
                    if (sizeof($attribute[1][0]) == 0)
                        $control_content = substr_replace($control_content, $control_name . '___id', $main_tag[0][1] + $attribute[1][1], 0);
                }
                else
                    $control_content = preg_replace('/(< *[^ >]+)/', '$1 id="' . $control_name . '___id"', $control_content, 1);
            }
            
            if (preg_match('/< *script[^>]*>/', $control_content, $script_start, PREG_OFFSET_CAPTURE))
            {
                $html = substr($control_content, 0, $script_start[0][1]);
                
                if (preg_match('/< *\/ *script[^>]*>/', $control_content, $script_end, PREG_OFFSET_CAPTURE))
                    $end = $script_end[0][1];
                else
                    $end = strlen($control_content);
                
                $start = $script_start[0][1] + strlen($script_start[0][0]);
                $code = substr($control_content, $start, $end - $start);

                if (preg_match_all('/function[ \n]+' . preg_quote(self::PREFIX) . '([a-zA-Z0-9_]+) ?\(([^\)]*)\)/', $code, $functions_matches, PREG_OFFSET_CAPTURE))
                {
                    $define_block .= '/* ' . $control_name . '*/' . self::NEW_LINE;
                    $functions_count = sizeof($functions_matches[0]);

                    for($i = 0; $i < $functions_count; $i++)
                    {
                        $define_block .= '$.neatwiDefineFunction("' . $functions_matches[1][$i][0] . '",';
                        $define_block .= '"' . $control_name . '",';
                        $define_block .= 'function(' . $functions_matches[2][$i][0] . ')';

                        if ($i == $functions_count - 1)
                            $function_body = trim(substr($code, $functions_matches[0][$i][1] + strlen($functions_matches[0][$i][0])));
                        else
                        {
                            $fun_start = $functions_matches[0][$i][1] + strlen($functions_matches[0][$i][0]);
                            $function_body = trim(substr($code, $fun_start, $functions_matches[0][$i+1][1] - $fun_start));
                        }

                        if ($function_body[strlen($function_body)-1] == ';')
                                $function_body = substr($function_body, 0, strlen($function_body)-1);
                        $define_block .= $function_body . ');' . self::NEW_LINE;
                    }
                }
            }
            else
                $html = $control_content;
            
            $controls_block .= addslashes($html) . "\"";
        }
        
        $controls_block .= "});";
        $controls_block = preg_replace('/\s\s+/', '', $controls_block);
        $define_block .= self::NEW_LINE . self::NEW_LINE;

        file_put_contents($file_name, '$(document).ready(function() { ' . self::NEW_LINE . self::NEW_LINE . $define_block . $controls_block . '});');
    }
}

?>

File: cs/NeatWI.cs
/*!
 * NeatWI 1.0
 *
 * http://www.neatwi.com
 *
 * Copyright 2014 Rafał Krawczyk
 * Released under the MIT license:
 *   http://www.opensource.org/licenses/mit-license.php
 */

using System.Globalization;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Web;

namespace NeatWI
{
    public class ControlsGenerator
    {
        #region Config

        // Directory of controls files
        const string SOURCE_DIRECTORY = "Controls";

        // Destination directory for js files
        const string DESTINATION_DIRECTORY = "Scripts";

        // Controls function names prefix
        const string PREFIX = "$";

        // If use controls categories
        const bool CATEGORIES = false;

        // New line sign in js files
        const string NEW_LINE = "\r\n";

        #endregion

        #region Language support - for user implementation

        /// <summary>
        /// If you use language support you have to implement this function
        /// </summary>
        /// <param name="language">Language to load</param>
        private void LoadLanguage(string language)
        {
            // For user implementation
            Thread.CurrentThread.CurrentUICulture = new CultureInfo(language);
            Thread.CurrentThread.CurrentCulture = CultureInfo.CreateSpecificCulture(language);
        }

        /// <summary>
        /// If you use language support you can implement this function
        /// so generator can load current language after generation
        /// </summary>
        /// <returns>Current language</returns>
        private string GetLanguage()
        {
            // For user implementation
            return Thread.CurrentThread.CurrentCulture.Name;
        }

        #endregion

        #region Methods for generation

        /// <summary>
        /// Generates controls without language support
        /// </summary>
        public void Generate()
        {
            Generate(null);
        }

        /// <summary>
        /// Generates controls for specified language
        /// </summary>
        /// <param name="languages">Array of language codes or names. Ex. new string[]{"en", "pl"}</param>
        public void Generate(string[] languages)
        {
            if (languages != null && languages.Length > 0)
            {
                string currentLanguage = GetLanguage();

                foreach (string language in languages)
                    GenerateControls(language);

                if (!string.IsNullOrEmpty(currentLanguage))
                    LoadLanguage(currentLanguage);
            }
            else
                GenerateControls(null);
        }

        #endregion

        #region Private members

        /// <summary>
        /// Generate controls for language
        /// </summary>
        /// <param name="language">Language of controls or empty</param>
        private void GenerateControls(string language)
        {
            if (language != null)
                LoadLanguage(language);

            if (CATEGORIES)
            {
                string[] directories = Directory.GetDirectories(Path.Combine(HttpContext.Current.Request.PhysicalApplicationPath, SOURCE_DIRECTORY));
                foreach (string directory in directories)
                    GenerateDirectory(new DirectoryInfo(directory).Name, language);
            }
            else
                GenerateDirectory(string.Empty, language);
        }

        /// <summary>
        /// Generates controls for direcotry
        /// </summary>
        /// <param name="category">Name of the category</param>
        /// <param name="language">Language controls or empty</param>
        private void GenerateDirectory(string category, string language)
        {
            StringBuilder defineBlock = new StringBuilder();
            StringBuilder controlsBlock = new StringBuilder();
            string categoryFileName = Path.Combine(HttpContext.Current.Request.PhysicalApplicationPath, DESTINATION_DIRECTORY) + "/neatwi_" + (!string.IsNullOrEmpty(category) ? category : "controls") + (!string.IsNullOrEmpty(language) ? "_" + language : "") + ".js";
            string[] controlPaths = System.IO.Directory.GetFiles(Path.Combine(HttpContext.Current.Request.PhysicalApplicationPath, SOURCE_DIRECTORY, category), "*.aspx");

            controlsBlock.Append("$.extend($.controls, {");
            bool first = true;

            foreach (string controlPath in controlPaths)
            {
                if (first)
                    first = false;
                else
                    controlsBlock.Append(",");

                string controlName = Path.GetFileNameWithoutExtension(new FileInfo(controlPath).Name);

                controlsBlock.Append("\"" + controlName + "\":\"");

                StringWriter sw = new StringWriter();
                HttpContext.Current.Server.Execute(Path.Combine("~/", SOURCE_DIRECTORY, category, controlName + ".aspx"), sw);
                string controlContent = sw.ToString();
                sw.Close();

                Match mainTag = Regex.Match(controlContent, "< *[^>]+>");
                if (mainTag.Success)
                {
                    Match attribute = Regex.Match(mainTag.Groups[0].Value, "[ '\"]class *= *['\"]([^'\"]*)['\"]");
                    if (attribute.Success)
                    {
                        string classes = attribute.Groups[1].Value.Trim();
                        if (classes.Length > 0)
                        {
                            string[] split = classes.Split(' ');
                            if (split[0] != controlName)
                                controlContent = controlContent.Insert(mainTag.Groups[0].Index + attribute.Groups[1].Index, controlName + " ");
                        }
                        else
                            controlContent = controlContent.Insert(mainTag.Groups[0].Index + attribute.Groups[1].Index, controlName);
                    }
                    else
                        controlContent = new Regex("(< *[^ >]+)").Replace(controlContent, "$1 class=\"" + controlName + "\"", 1);

                    mainTag = Regex.Match(controlContent, "< *[^>]+>");
                    attribute = Regex.Match(mainTag.Groups[0].Value, "[ '\"]id *= *['\"]([^'\"]*)['\"]");
                    if (attribute.Success)
                    {
                        if (attribute.Groups[1].Length == 0)
                            controlContent = controlContent.Insert(mainTag.Groups[0].Index + attribute.Groups[1].Index, controlName + "___id");
                    }
                    else
                        controlContent = new Regex("(< *[^ >]+)").Replace(controlContent, "$1 id=\"" + controlName + "___id\"", 1);
                }

                string html;
                Match scriptStart = Regex.Match(controlContent, "< *script[^>]*>");

                if (scriptStart.Success)
                {
                    html = controlContent.Substring(0, scriptStart.Index);
                    int start = scriptStart.Index + scriptStart.Length;
                    int end;

                    Match scriptEnd = Regex.Match(controlContent, "< */ *script[^>]*>");
                    if (scriptEnd.Success)
                        end = scriptEnd.Index;
                    else
                        end = controlContent.Length;

                    string code = controlContent.Substring(start, end - start);

                    MatchCollection functions = Regex.Matches(code, @"function[ \n]+" + Regex.Escape(PREFIX) + @"([a-zA-Z0-9_]+) ?\(([^\\)]*)\)");
                    if (functions.Count > 0)
                    {
                        defineBlock.Append("/* " + controlName + "*/" + NEW_LINE);
                        int functionsCount = functions.Count;

                        for (int i = 0; i < functionsCount; i++)
                        {
                            Match function = functions[i];
                            defineBlock.Append("$.neatwiDefineFunction(\"" + function.Groups[1].Value + "\",");
                            defineBlock.Append("\"" + controlName + "\",");
                            defineBlock.Append("function(" + function.Groups[2].Value + ")");

                            string functionBody;

                            if (i == functionsCount - 1)
                                functionBody = code.Substring(function.Index + function.Length).Trim();
                            else
                            {
                                int functionStart = function.Index + function.Length;
                                functionBody = code.Substring(functionStart, functions[i + 1].Index - functionStart).Trim();
                            }

                            if (functionBody[functionBody.Length - 1] == ';')
                                functionBody = functionBody.Substring(0, functionBody.Length - 1);

                            defineBlock.Append(functionBody + ");" + NEW_LINE);
                        }
                    }

                }
                else
                    html = controlContent;

                controlsBlock.Append(html.Trim().Replace(System.Environment.NewLine, "").Replace("\\", "\\\\").Replace("\"", "\\\"").Replace("'", "\\'") + "\"");
            }

            controlsBlock.Append("});");
            defineBlock.Append(NEW_LINE + NEW_LINE);

            using (StreamWriter outFile = new StreamWriter(categoryFileName))
            {
                outFile.Write("$(document).ready(function() { " + NEW_LINE + NEW_LINE + defineBlock.ToString() + controlsBlock.ToString() + "});");
                outFile.Flush();
                outFile.Close();
            }
        }

        #endregion
    }
}
